### PR TITLE
Changing pointers to empty strings to nil pointers

### LINKED ISSUE #908: fake rp writes back the internal representation of the manifest
For example, the external request includes
```yaml
plan: {}
```
but after an update succeeds, the manifest is updated with the internal representation
```yaml
plan:
  name: ""
  product: ""
  promotionCode: ""
  publisher: ""
```
This was caused by https://github.com/openshift/openshift-azure/pull/821 where we updated the fake RP to use the internal representation of OpenShiftManagedCluster in order to handle multiple versions.

The actual problem is in `pkg/api/convertto20180930preview.go`. We should avoid returning pointers to empty strings. The same fix should be applied in `pkg/api/converttoadmin.go`.
/assign Makdaam 
/kind bug

### PR DESCRIPTION
**Pull request by Makdaam**
Fixes #908.

Added a small func to replace pointers to "" strings with nil. Wrapped all string pointers in converttoadmin.go and converttov20180930preview.go with it.

### CODE REVIEW

**Comment by Makdaam** [comment]
misclick

File: pkg/api/util.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package api
+
+// nilIfStringEmpty returns a nil if the string is 0 length and the string pointer otherwise
+func nilIfStringEmpty(s *string) *string {
+	if len(*s) == 0 {
```
Comment (kargakis) [inline pkg/api/util.go]: To be on the safe side:
```go
if s == nil || len(*s) == 0 {
    return nil
}
```

**Comment by kargakis** [comment]
Can you also add a unit test that proves #908 is fixed?

File: pkg/api/util.go (hunk 1 above)
Comment (Makdaam) [reply pkg/api/util.go]: Thanks for catching this.

**Comment by Makdaam** [comment]
@kargakis added unit tests for empty string to nil conversion.

**Comment by kargakis** [comment]
lgtm, please squash to a single commit and I can tag it

**Comment by Makdaam** [comment]
@kargakis squashed

File: pkg/api/convertertov20180930preview.go [hunk 2]
```diff
@@ -8,44 +8,44 @@ import (
...
-		Publisher:     &cs.Plan.Publisher,
+		Name:          nilIfStringEmpty(&cs.Plan.Name),
+		Product:       nilIfStringEmpty(&cs.Plan.Product),
+		PromotionCode: nilIfStringEmpty(&cs.Plan.PromotionCode),
+		Publisher:     nilIfStringEmpty(&cs.Plan.Publisher),
 	}
 
 	provisioningState := v20180930preview.ProvisioningState(cs.Properties.ProvisioningState)
 	oc.Properties = &v20180930preview.Properties{
 		ProvisioningState: &provisioningState,
```
Comment (kargakis) [inline pkg/api/convertertov20180930preview.go]: Hrm, I think provisioningState, count, vmSize, osType, and role also need to be handled the same way?

Comment (Makdaam) [reply pkg/api/convertertov20180930preview.go]: Amended the commit with provisioningState, vmSize, osType and role *""→nil conversion. Also changed the test case a bit to include internal fields of AgentPoolProfile.

Comment (Makdaam) [reply pkg/api/convertertov20180930preview.go]: nvm, forgot about converttoadmin.go and _test. BRB with more amends.

Comment (Makdaam) [reply pkg/api/convertertov20180930preview.go]: amended with converttoadmin.go changes

**Comment by Makdaam** [comment]
After discussion a different approach to fixing this issue has been chosen: 
- store the plan internally as pointers and pass it on as is
- don't modify other fields as the difference between an empty value and NULL value would be lost in conversion

**Pull request closed without merging**